### PR TITLE
Optimizing unsigned DIV and REM ops

### PR DESCRIPTION
**Pull request by arunetm**
Replacing unsigned DIV and REM ops with MUL, ADD, and shifts.

### CODE REVIEW

**Comment by arunetm** [comment]
Combined impact of current PR and https://github.com/Microsoft/ChakraCore/pull/4438 

```
On Atom: (CHT)
Running jetstream...

TEST - NATIVE                BASE            TEST         DIFF   RATIO(%)
---------------------------------------------------------------------------
bigfib.cpp                 25.0 +-0.0%    30.0 +-0.0%      5.0   20.0% <-IMPROVED
container.cpp              25.0 +-0.0%    25.0 +-0.0%      0.0    0.0%
dry.c                      15.0 +-0.0%    15.5 +-3.2%      0.5    3.3%
float-mm.c                 28.0 +-0.0%    31.0 +-0.0%      3.0   10.7% <-IMPROVED
gcc-loops.cpp              30.0 +-0.0%    30.0 +-0.0%      0.0    0.0%
hash-map                    9.0 +-0.0%     9.0 +-0.0%      0.0    0.0%
n-body.c                   13.0 +-0.0%    13.0 +-0.0%      0.0    0.0%
quicksort.c                28.0 +-0.0%    31.0 +-0.0%      3.0   10.7% <-IMPROVED
towers.c                   16.0 +-0.0%    16.0 +-0.0%      0.0    0.0%
cdjs                       10.0 +-0.0%    10.0 +-0.0%      0.0    0.0%
---------------------------------------------------------------------------
MEAN SCORE                 18.3 +-0.0%    19.0 +-0.0%      0.8    4.3% <-IMPROVED
```

**Comment by arunetm** [comment]
@dotnet-bot test Ubuntu shared_ubuntu_linux_test please

**Comment by arunetm** [comment]
ping.

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -24898,17 +24898,25 @@ Lowerer::LowerDivI4(IR::Instr * instr)
...
-        InsertTestBranch(nominatorOpnd, IR::IntConstOpnd::New(pow2 - 1, TyInt32, this->m_func, true),
-            Js::OpCode::BrNeq_A, bailOutLabel, nonBailOutInstr);
-        nonBailOutInstr->m_opcode = Js::OpCode::Shr_A;
-        nonBailOutInstr->ReplaceSrc2(IR::IntConstOpnd::New(Math::Log2(pow2), TyInt32, this->m_func, true));
-        LowererMD::ChangeToShift(nonBailOutInstr, false);
-        LowererMD::Legalize(nonBailOutInstr);
+        if (Math::IsPow2(denominatorOpnd->AsIntConstOpnd()->AsInt32()))
+        {
+            int pow2 = denominatorOpnd->AsIntConstOpnd()->AsInt32();
+            InsertTestBranch(nominatorOpnd, IR::IntConstOpnd::New(pow2 - 1, TyInt32, this->m_func, true),
```
Comment (LouisLaf) [inline lib/Backend/Lower.cpp]: Since pow2 is user controlled, these IntConstOpnd shouldn't be marked as dontEncode.

Comment (LouisLaf) [reply lib/Backend/Lower.cpp]: The rest looks good.  Nice wins!

Comment (arunetm) [reply lib/Backend/Lower.cpp]: Sure, removing the dontEncode flag.

**Comment by arunetm** [comment]
@dotnet-bot test OSX static_osx_osx_release please

**Comment by arunetm** [comment]
squashing commits and rebasing with master.

**Pull request merged**